internal/ratelimit/config.go

```diff
@@ -28,3 +41,124 @@ type Config struct {
 
 // Configs is an ordered set of Config.
 type Configs []*Config
+
+// Limits creates a slice of rate.Limit from the Configs. This will enumerate
+// every combination of resource+action, defining a Limit for each.
+func (c Configs) Limits() ([]*rate.Limit, error) {
+	const op = "ratelimit.(Configs).Limits"
+
+	defaults := make(map[string]*rate.Limit, len(resource.Map)*len(action.Map))
+
+	allResources := make([]resource.Type, 0, len(resource.Map))
+	for _, res := range resource.Map {
+		switch res {
+		case resource.Unknown, resource.All, resource.Controller:
+			continue
+		}
+		allResources = append(allResources, res)
+	}
+
+	for _, res := range allResources {
+		validActions, err := action.SetForResource(res)
+		if err != nil {
+			// TODO return an error instead?
+			panic(fmt.Sprintf("No actions registered for resource %q", res.String()))
+		}
+
+		for _, a := range validActions {
+			key := fmt.Sprintf("%s:%s:%s", res.String(), a.String(), rate.LimitPerTotal)
+			switch a {
+			case action.List:
+				defaults[key] = &rate.Limit{
+					Resource:    res.String(),
+					Action:      a.String(),
+					Per:         rate.LimitPerTotal,
+					Unlimited:   false,
+					MaxRequests: DefaultListRequestLimit,
+					Period:      DefaultListPeriod,
+				}
+			default:
+				defaults[key] = &rate.Limit{
+					Resource:    res.String(),
+					Action:      a.String(),
+					Per:         rate.LimitPerTotal,
+					Unlimited:   false,
+					MaxRequests: DefaultRequestLimit,
+					Period:      DefaultPeriod,
+				}
+			}
+		}
+	}
+
+	for _, cc := range c {
+		var resourceSet []resource.Type
+		switch {
+		case len(cc.Resources) == 1 && cc.Resources[0] == "*":
+			resourceSet = allResources
+		default:
+			for _, r := range cc.Resources {
+				rr, ok := resource.Map[r]
+				if !ok {
+					return nil, fmt.Errorf("unknown resource %s", r)
```

Comment:
should we define a new error code for this?

Reply:
That seems like a good idea. In looking through the existing codes, I am a little surprised that we don't have something for a configuration error, which is what I was thinking this would be. It seems that for most of the config parsing we don't return errors with error codes.

Do you think I should make an error code specific to specifying a resource that is not known? Or should I look to add a more generic "configuration error"?

Reply:
I'd start with `configuration error` in this case.